### PR TITLE
Ex4

### PR DESCRIPTION
**Pull request by zyahav**

### CODE REVIEW

File: ex4B/ex4Spec.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+var ex4 = require('./ex4.js');
+
+describe("toNumber", function() {
+
+	it("converte the chacter 7 & 1 string  to a number 71", function() {
```
Comment (myok12) [inline ex4B/ex4Spec.js]: Please add a few more test cases: single digit (5), negative number (-3), decimal (0.6), negative and decimal (-3.5), many digits (12351231), etc. And their negatives: a5, 5g, 3-, --3, -3-, .4, 4.4.4, 4., aa, etc.

Comment (myok12) [inline ex4B/ex4Spec.js]: Converts 2 digits

File: ex4B/ex4Spec.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+		expect(ex4.toNumber("71")).toMatch(71);
+	});
+
+	it("return null, string doesn't contain a valid number", function() {
```
Comment (myok12) [inline ex4B/ex4Spec.js]: returns null when the input doesn't contain a valid number

Comment (zyahav) [inline ex4B/ex4Spec.js]: done

File: ex4B/ex4Spec.js (hunk 1 above)
Comment (zyahav) [inline ex4B/ex4Spec.js]: "-3" shoud it return -3 ???

Comment (zyahav) [inline ex4B/ex4Spec.js]: toNumber return only number ? +numbers

Comment (myok12) [inline ex4B/ex4Spec.js]: Normally, when people say numbers, they mean decimal numbers (https://en.wikipedia.org/wiki/Decimal). So, yes, unless described differently, you should support negative numbers (-3) and decimals with a decimal point (3.14) and, of course, both (-3.14). 
There are many other number notations, but this it the "peoples" default one.

File: ex4B/ex4.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * toNumber("71") === 71
+ * toNumber("a7") === null
+ */
+function toNumber(str) {
+
+	var tmpStr = '';
+
+	for (var i = 0; i < str.length; i++) {
+
+		if (str.slice(i, i + 1) >= 0 && str.slice(i, i + 1) <= 9) {
```
Comment (myok12) [inline ex4B/ex4.js]: Are you sure this works? Do your test run green?

Comment (zyahav) [inline ex4B/ex4.js]: all test are green

File: ex4B/ex4.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * For example:
+ * toNumber("71") === 71
+ * toNumber("a7") === null
+ */
+function toNumber(str) {
+
+	var tmpStr = '', i = 0, dotRepeat = null;
+
+	// Negative sine can appear only at the start of string and followed by number
+	if (str.slice(i, i + 1) === '-' && isNumber(str.slice(i+1, i + 2))) {
```
Comment (myok12) [inline ex4B/ex4.js]: Using `i` here rather than `0` or `1` doesn't buy you anything. It just makes the code harder to understand. Switch to `0` and `1`.

Comment (myok12) [inline ex4B/ex4.js]: When you want to look at a single char in a string use, `str.charAt(i)` instead of `str.slice(i, i+1).

File: ex4B/ex4.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	
+		tmpStr += str.slice(i, i + 2);  
+		i = 2;
+	}
+
+	// Dot cannot appear at the start or and of string
```
Comment (myok12) [inline ex4B/ex4.js]: or ==> or end

File: ex4B/ex4.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	if (str[0] === '.' || str[str.length - 1]=== '.') {
+		return null;
+	}
+
+	for (;i < str.length; i++) {
+
+		if (isNumber(str.slice(i, i + 1)) ||  str.slice(i, i + 1) === '.' && !dotRepeat) {
+ 
+			// Ziro tolerants for dots
```
Comment (myok12) [inline ex4B/ex4.js]: Ziro is not a word. Zero is.

File: ex4B/ex4.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * For example:
+ * toNumber("71") === 71
+ * toNumber("a7") === null
+ */
+function toNumber(str) {
+
+	var tmpStr = '', i = 0, dotRepeat = null;
+
+	// Negative sine can appear only at the start of string and followed by number
```
Comment (myok12) [inline ex4B/ex4.js]: sine --> sign

File: ex4B/ex4.js (hunk 4 above)
Comment (myok12) [inline ex4B/ex4.js]: if you end up not using i here, move the initialization (i=0) to the for loop.

File: ex4B/ex4.js (hunk 6 above)
Comment (myok12) [inline ex4B/ex4.js]: tolerants --> tolerance.

File: ex4B/ex4.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+ * Write a function that convers a string into number; return null if the string
+ * doesn't contain a valid number.
+ *
+ * For example:
+ * toNumber("71") === 71
+ * toNumber("a7") === null
+ */
+function toNumber(str) {
+
+	var tmpStr = '', i = 0, dotRepeat = null;
```
Comment (myok12) [inline ex4B/ex4.js]: if `dotRepeats` acts as a boolean (it should), initialize it to `false` rather than `null` (even though they work the same).

Comment (myok12) [inline ex4B/ex4.js]: Also, `dotRepeat` sounds like there's an option to allow several dots, which is not the case. How about renaming the variable to `dotAppeared` (`false` at first, `true` after finding a dot -- same as you have now) or `allowDot` (`true` at first, `false` after finding a dot -- inverse of what you have now).

File: ex4B/ex4.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+			tmpStr += str.slice(i, i + 1);  
+
+		} else {
+
+			return null;
+
+		}
+	}
+
+	return Number(tmpStr);
```
Comment (myok12) [inline ex4B/ex4.js]: Ok. This works, but you shouldn't use Number. In this question I want YOU to implement this yourself.

File: ex4B/ex4Spec.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+		expect(toNumber("71")).toMatch(71);
+	});
+
+	it("converts many digits", function() {
+		expect(toNumber("712387328182")).toMatch(712387328182);
+	});
+	
+	///////////////////////////////////////////////////////////////////////	
+	// Numbers negative
+	it("converts a single negative digit", function() {
```
Comment (myok12) [inline ex4B/ex4Spec.js]: single digit is 3, negative number is -17581. You probably meant `a negative number with a single digit`. Shortening the sentence makes it incorrect.

Comment (myok12) [inline ex4B/ex4Spec.js]: You could say `a negative single-digit number`

File: ex4B/ex4Spec.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+		expect(toNumber("-3")).toBe(-3);
+	});
+
+	it("converts a negative 2 digits", function() {
```
Comment (myok12) [inline ex4B/ex4Spec.js]: `a negative 2-digit number`

File: ex4B/ex4Spec.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+		expect(toNumber("-32")).toBe(-32);
+	});
+
+	it("converts many netgative digits", function() {
+		expect(toNumber("-712387328182")).toMatch(712387328182);
```
Comment (myok12) [inline ex4B/ex4Spec.js]: Your test should be `-712387328182`

File: ex4B/ex4Spec.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+	it("return null, 2 digits. negative sine cannot be last", function() {
+		expect(toNumber("32-")).toBeNull();
+	});
+
+	it("return null, many digits. negative sine cannot be last ", function() {
+		expect(toNumber("712387328182-")).toBeNull();
+	});
+
+	///////////////////////////////////////////////////////////////////////
```
Comment (myok12) [inline ex4B/ex4Spec.js]: What about negative sign in the middle: `751-123`?

File: ex4B/ex4Spec.js [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+	it("return null, decimal and negative. negative and dot cannot be the first sine", function() {
+		expect(toNumber("-.6")).toBeNull();
+	});
+
+	it("return null, decimal and negative 2 digits. negative and dot cannot be the first sine", function() {
+		expect(toNumber("-.62")).toBeNull();
+	});
+
+	it("creturn null, decimal and negative many digits. negative and dot cannot be the first sine", function() {
```
Comment (myok12) [inline ex4B/ex4Spec.js]: What about `.-41`

File: ex4B/ex4Spec.js [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+	it("returns null, input single number and a single char", function() {
+		expect(toNumber("7a")).toBeNull();
+	});
+
+	it("returns null, input many single characters", function() {
+		expect(toNumber("akjadiaoihdsaoi")).toBeNull();
+	});
+
+	///////////////////////////////////////////////////////////////////////
```
Comment (myok12) [inline ex4B/ex4Spec.js]: What about ``?

File: ex4B/ex4.js (hunk 7 above)
Comment (zyahav) [inline ex4B/ex4.js]: Done

Comment (zyahav) [inline ex4B/ex4.js]: Done

File: ex4B/ex4.js (hunk 6 above)
Comment (zyahav) [inline ex4B/ex4.js]: Done

File: ex4B/ex4.js (hunk 4 above)
Comment (zyahav) [inline ex4B/ex4.js]: Done

File: ex4B/ex4.js [hunk 16]
```diff
@@ -8,9 +8,9 @@
  */
 function toNumber(str) {
 
-	var tmpStr = '', i = 0, dotRepeat = null;
+	var tmpStr = '', i = 0, allowDot = true;
 
-	// Negative sine can appear only at the start of string and followed by number
+	// Negative sige can appear only at the start of string and followed by number
```
Comment (myok12) [inline ex4B/ex4.js]: sige --> sign

File: ex4B/ex4Spec.js (hunk 11 above)
Comment (zyahav) [inline ex4B/ex4Spec.js]: Done

File: ex4B/ex4Spec.js (hunk 12 above)
Comment (zyahav) [inline ex4B/ex4Spec.js]: Replace toMatch to toBe
done

File: ex4B/ex4Spec.js (hunk 13 above)
Comment (zyahav) [inline ex4B/ex4Spec.js]: Done

File: ex4B/ex4Spec.js [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+describe("function toNumber", function() {
+
+	///////////////////////////////////////////////////////////////////////	
+	// Numbers
+	it("converts single digit", function() {
+		expect(toNumber("5")).toBe(5);
+	});
+	
+	it("converts 2 digits", function() {
+		expect(toNumber("71")).toMatch(71);
```
Comment (myok12) [inline ex4B/ex4Spec.js]: toEqual

File: ex4B/ex4.js (hunk 9 above)
Comment (zyahav) [inline ex4B/ex4.js]: Done

File: ex4B/ex4Spec.js (hunk 17 above)
Comment (zyahav) [inline ex4B/ex4Spec.js]: Done

File: ex4B/ex4.js [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+function toNumber(str) {
+
+	var strNum = '', strDec = '', number = null, i = 0, allowDot = true, negative = false;
+
+	// Negative sige can appear only at the start of string and followed by number
+	//if (str.slice(0, 1) === '-' && isNumber(str.slice(1, 2))) {
+	
+	if (str.charAt(0) === '-' ) {
+		
+		if (isNumber(str.charAt(1))) {
```
Comment (myok12) [inline ex4B/ex4.js]: Isn't this checking the same as in line 30 below?
Actually it probably doesn't work because of the `str[0]` below (which needs to change between 0 or 1 depending on the minus sign.

File: ex4B/ex4.js [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+		if (isNumber(str.charAt(i)) ||  str.charAt(i) === '.' && allowDot) {
+ 
+			// Zero tolerance for dots
+			if (str.charAt(i) === '.') {
+
+				allowDot = false;
+				continue;
+			}
+
+			if (!allowDot) {
```
Comment (myok12) [inline ex4B/ex4.js]: Now that you use `allowDot` to "know" whether it's the integer part or the decimal part, the name definitely doesn't explain that. I would rename to `seenDot` (for which you'll need to negate its conditions).

File: ex4B/ex4.js [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+	return number;
+}
+
+function strToNumber(str) {
+
+	var num = 0;
+
+	for (var i = 0; i < str.length; i++) { 
+		num = num * 10;
+		num += str[i].charCodeAt() - 48;
```
Comment (myok12) [inline ex4B/ex4.js]: 48 isn't very descriptive. Change this to "0".charCodeAt(). Or even better, create a constant (search javascript const) at the top which will contain this value (otherwise, it will waste CPU every time it needs the value in the loop).

File: ex4B/ex4.js (hunk 18 above)
Comment (zyahav) [inline ex4B/ex4.js]: will talk about it

File: ex4B/ex4.js (hunk 19 above)
Comment (zyahav) [inline ex4B/ex4.js]: Done :Replace allowDot ==> seenDot

File: ex4B/ex4.js (hunk 18 above)
Comment (zyahav) [inline ex4B/ex4.js]: Done

File: ex4B/ex4.js (hunk 20 above)
Comment (zyahav) [inline ex4B/ex4.js]: Done: raplace -48 to const

File: ex4B/ex4.js [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+ * doesn't contain a valid number.
+ *
+ * For example:
+ * toNumber("71") === 71
+ * toNumber("a7") === null
+ */
+function toNumber(str) {
+
+	var strNum = '', strDec = '', number = null, i = 0, seenDot = false, negative = false;
+	const CODE_FOR_0 = "0".charCodeAt();
```
Comment (myok12) [inline ex4B/ex4.js]: If you move the const outside of the function, you will not need to pass it it the other functions (it's within the closure of the page).

Think of it this way: You modified `strToNumber(str) {...}` `strToNumber(str, CODE_FOR_0) {...}`. Does strToNumber really needs CODE_FOR_0 to do its work? No, because you could have calculated it again if you really wanted to. So, it shouldn't be in the signature of this method, hence it's better to rely on a more-"global" const.

File: ex4B/ex4.js [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+	
+	if (str.charAt(0) === '-' ) {
+		
+		if (isNumber(str.charAt(1))) {
+
+			negative = true;  
+			i = 1;
+
+		} else {
+
```
Comment (myok12) [inline ex4B/ex4.js]: Even though we discussed it, I'd put a comment here saying: `// Will catch "-.3", for example.

Comment (zyahav) [inline ex4B/ex4.js]: Done

File: ex4B/ex4.js (hunk 21 above)
Comment (zyahav) [inline ex4B/ex4.js]: Done

File: ex4B/ex4.js [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+	var num = 0;
+
+	for (var i = str.length; i > 0; i--) { 
+		num = num  / 10;
+		num += str[i-1].charCodeAt() - CODE_FOR_0;
+	}
+	return num;
+}
+
+function isNumber(val) {
```
Comment (myok12) [inline ex4B/ex4.js]: Should be named isDigit (digit is one character number. a number can contain more than one digit)

File: ex4B/ex4.js [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * For example:
+ * toNumber("71") === 71
+ * toNumber("a7") === null
+ */
+
+const CODE_FOR_0 = "0".charCodeAt();
+
+function toNumber(str) {
+	var strNum = '', strDec = '', number = null, i = 0, seenDot = false, negative = false;
```
Comment (myok12) [inline ex4B/ex4.js]: Get into the habit of only declaring the `var`s when needed -- this adds to clarify (because you know what section of the code pertains to what variable, and also it helps seeing what the initial value is in the right context).
